File: blue-core/src/blue/orchestra/blueSynthBuilder/Preset.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2004 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.orchestra.blueSynthBuilder;

import electric.xml.Element;
import electric.xml.Elements;
import java.io.Serializable;
import java.rmi.dgc.VMID;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;

/**
 * @author steven
 */
public class Preset implements Serializable, Comparable<Preset> {

    private String presetName = "";

    private HashMap valuesMap = new HashMap();

    private String uniqueId;

    public Preset() {
        this.uniqueId = Integer.toString(new VMID().hashCode());
    }

    public static Preset createPreset(BSBGraphicInterface bsbInterface) {
        Preset preset = new Preset();

        for (Iterator iter = bsbInterface.iterator(); iter.hasNext();) {
            BSBObject bsbObj = (BSBObject) iter.next();

            String objectName = bsbObj.getObjectName();

            if (objectName == null || objectName.length() == 0) {
                continue;
            }

            String val = bsbObj.getPresetValue();

            if (val != null) {
                preset.valuesMap.put(objectName, val);
            }

        }

        return preset;
    }

    public void updatePresets(BSBGraphicInterface bsbInterface) {

        valuesMap.clear();

        for (Iterator iter = bsbInterface.iterator(); iter.hasNext();) {
            BSBObject bsbObj = (BSBObject) iter.next();

            String objectName = bsbObj.getObjectName();

            if (objectName == null || objectName.length() == 0) {
                continue;
            }

            String val = bsbObj.getPresetValue();

            if (val != null) {
                valuesMap.put(objectName, val);
            }

        }
    }

    public static Preset loadFromXML(Element data) {
        Preset p = new Preset();

        p.setPresetName(data.getAttributeValue("name"));

        String uniqueId = data.getAttributeValue("uniqueId");
        if (uniqueId != null && uniqueId.length() > 0) {
            p.uniqueId = uniqueId;
        }

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();

            if (node.getName().equals("setting")) {
                String name = node.getAttributeValue("name");
                String val = node.getTextString();

                p.addSetting(name, val);
            }
        }

        return p;

    }

    public Element saveAsXML() {
        Element retVal = new Element("preset");
        retVal.setAttribute("name", presetName);
        retVal.setAttribute("uniqueId", uniqueId);

        Object[] keys = valuesMap.keySet().toArray();

        Arrays.sort(keys);

        for (int i = 0; i < keys.length; i++) {
            String key = (String) keys[i];
            String val = (String) valuesMap.get(key);

            retVal.addElement("setting").setAttribute("name", key).setText(val);
        }

        return retVal;
    }

    public String getUniqueId() {
        return uniqueId;
    }

    public void addSetting(String objectName, String objectVal) {
        valuesMap.put(objectName, objectVal);
    }

    public String getSetting(String objectName) {
        return (String) valuesMap.get(objectName);
    }

    /**
     * @return Returns the preset name.
     */
    public String getPresetName() {
        return presetName;
    }

    /**
     * @param name
     *            The preset name to set.
     */
    public void setPresetName(String name) {
        this.presetName = name;
    }

    @Override
    public String toString() {
        return getPresetName();
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Comparable#compareTo(java.lang.Object)
     */
    public int compareTo(Preset b) {
        return this.getPresetName().compareTo(b.getPresetName());
    }

    /**
     * @param graphicInterface
     */
    public void setInterfaceValues(BSBGraphicInterface graphicInterface) {
        for (Iterator iter = graphicInterface.iterator(); iter.hasNext();) {
            BSBObject bsbObj = (BSBObject) iter.next();
            String objName = bsbObj.getObjectName();

            String setting = getSetting(objName);

            if (setting != null) {
                bsbObj.setPresetValue(setting);
            }

        }
    }

    public void synchronizeWithInterface(BSBGraphicInterface graphicInterface) {
        HashMap nameMap = new HashMap();

        for (Iterator iter = graphicInterface.iterator(); iter.hasNext();) {
            BSBObject bsbObj = (BSBObject) iter.next();
            String objName = bsbObj.getObjectName();

            nameMap.put(objName, objName);
        }

        // clear out any existing values that do not have a widget in interface
        for (Iterator iter = this.valuesMap.keySet().iterator(); iter.hasNext();) {
            String key = (String) iter.next();

            if (!nameMap.containsKey(key)) {
                System.out.println("Removing preset with objectName: " + key);
                iter.remove();
            }
        }

        // add values that are in interface but not in preset
        for (Iterator iter = graphicInterface.iterator(); iter.hasNext();) {
            BSBObject bsbObj = (BSBObject) iter.next();
            String objName = bsbObj.getObjectName();

            if (objName == null || objName.length() == 0) {
                continue;
            }

            if (!valuesMap.containsKey(objName)) {

                String val = bsbObj.getPresetValue();

                if (val != null) {
                    System.out.println("adding preset for objectName: "
                            + objName);
                    valuesMap.put(objName, val);
                }
            }
        }
    }
}

File: packages/blue-data/src/instruments/blue-synth-builder/preset.ts
import { Element } from '../../serialization/xml-reader';
import { generatePrefixedUuid } from '../../utilities/uuid';
import type { BSBGraphicInterface } from './bsb-graphic-interface';
import type { BSBWidget } from './bsb-widget';
import { BSBGroup } from './bsb-group';

export class Preset {
  presetName = '';
  uniqueId = generatePrefixedUuid('preset');
  private _valuesMap = new Map<string, string>();

  getPresetName(): string {
    return this.presetName;
  }

  setPresetName(name: string): void {
    this.presetName = name;
  }

  getUniqueId(): string {
    return this.uniqueId;
  }

  getValuesMap(): Map<string, string> {
    return new Map(this._valuesMap);
  }

  setValuesMap(values: Map<string, string>): void {
    this._valuesMap = new Map(values);
  }

  getValue(objectName: string): string | undefined {
    return this._valuesMap.get(objectName);
  }

  setValue(objectName: string, value: string): void {
    this._valuesMap.set(objectName, value);
  }

  updatePresets(gInterface: any): void {
    this._valuesMap.clear();
    const visit = (widgets: any[]): void => {
      for (const widget of widgets) {
        if (widget.objectName) {
          const value =
            typeof widget.getPresetValue === 'function'
              ? widget.getPresetValue()
              : widget.value !== undefined
                ? String(widget.value)
                : null;
          if (value !== null && value !== undefined) {
            this._valuesMap.set(widget.objectName, value);
          }
        }
        if (widget.getChildren && typeof widget.getChildren === 'function') {
          visit(widget.getChildren());
        }
      }
    };
    visit(gInterface.getRootGroup().getChildren());
  }

  /**
   * Synchronize this preset's stored values with the current BSB interface.
   *
   * Mirrors Java `Preset.synchronizeWithInterface(BSBGraphicInterface)`:
   * 1. Remove preset values whose widget objectName no longer exists in the interface.
   * 2. Add default preset values for widgets that exist in the interface but
   *    are not yet stored in this preset.
   */
  synchronizeWithInterface(graphicInterface: BSBGraphicInterface): void {
    // Build a set of all widget objectNames currently in the interface
    const nameSet = new Set<string>();
    const collectNames = (widgets: BSBWidget[]): void => {
      for (const widget of widgets) {
        if (widget.objectName) {
          nameSet.add(widget.objectName);
        }
        if (widget instanceof BSBGroup) {
          collectNames(widget.getChildren());
        }
      }
    };
    collectNames(graphicInterface.getRootGroup().getChildren());

    // Remove preset entries whose objectName no longer exists in the interface
    for (const key of [...this._valuesMap.keys()]) {
      if (!nameSet.has(key)) {
        this._valuesMap.delete(key);
      }
    }

    // Add default values for widgets present in the interface but missing from the preset
    const addMissing = (widgets: BSBWidget[]): void => {
      for (const widget of widgets) {
        const objName = widget.objectName;
        if (objName && objName.length > 0 && !this._valuesMap.has(objName)) {
          const val = widget.getPresetValue();
          if (val !== null && val !== undefined) {
            this._valuesMap.set(objName, val);
          }
        }
        if (widget instanceof BSBGroup) {
          addMissing(widget.getChildren());
        }
      }
    };
    addMissing(graphicInterface.getRootGroup().getChildren());
  }

  saveAsXML(): Element {
    const elem = new Element('preset');
    elem.setAttribute('name', this.presetName);
    elem.setAttribute('uniqueId', this.uniqueId);
    const sortedKeys = [...this._valuesMap.keys()].sort();
    for (const key of sortedKeys) {
      const setting = elem.addElement('setting');
      setting.setAttribute('name', key);
      setting.setText(this._valuesMap.get(key) ?? '');
    }
    return elem;
  }

  static loadFromXML(data: Element): Preset {
    const preset = new Preset();
    preset.presetName = data.getAttribute('name') ?? '';
    const uniqueId = data.getAttribute('uniqueId');
    if (uniqueId) {
      preset.uniqueId = uniqueId;
    }
    const settings = data.getElements('setting');
    while (settings.hasMoreElements()) {
      const setting = settings.next();
      const name = setting.getAttribute('name');
      const value = setting.getTextString();
      if (name) {
        preset._valuesMap.set(name, value ?? '');
      }
    }
    return preset;
  }

  deepCopy(presetIdMap?: Map<string, string>): Preset {
    const copy = new Preset();
    copy.presetName = this.presetName;
    if (this.uniqueId) {
      presetIdMap?.set(this.uniqueId, copy.uniqueId);
    }
    copy._valuesMap = new Map(this._valuesMap);
    return copy;
  }
}
